### PR TITLE
arch: x86: enhance number of cpu supports in smp

### PR DESCRIPTION
**Pull request by najumon1980**
Remove the limitation for number of cpu support in smp for x86 architecture. This changes remove x86 limitation for maximum number of CPUs supported and also update kernel limitation to 32 CPU instead of 8.

### CODE REVIEW

**Comment by najumon1980** [comment]
please remove stale label

**Comment by nashif** [comment]
> please remove stale label

you can remove it yourself.

File: arch/x86/include/intel64/kernel_arch_data.h [hunk 1]
```diff
@@ -26,6 +26,7 @@ struct x86_cpuboot {
 	size_t stack_size;	/* size of stack */
 	arch_cpustart_t fn;	/* kernel entry function */
 	void *arg;		/* argument for above function */
+	uint8_t cpu_id;		/* APIC ID */
```
Comment (dcpleung) [inline arch/x86/include/intel64/kernel_arch_data.h:29]: Shouldn't this be guarded under `CONFIG_ACPI`?

File: tests/arch/x86/nmi/src/main.c [hunk 2]
```diff
@@ -17,13 +17,13 @@
 
 static volatile int int_handler_executed;
 
-extern uint8_t z_x86_nmi_stack[];
+extern uint8_t z_x86_nmi_stack0[];
```
Comment (dcpleung) [inline tests/arch/x86/nmi/src/main.c:20]: The changes here are not in-line with the commit message. Could you re-phrase the commit message?

File: arch/x86/core/intel64/cpu.c [hunk 3]
```diff
@@ -15,133 +17,37 @@
...
 };
 
-/*
- * Send the INIT/STARTUP IPI sequence required to start up CPU 'cpu_num', which
- * will enter the kernel at fn(arg), running on the specified stack.
- */
-
-void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
-		    arch_cpustart_t fn, void *arg)
+void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz, arch_cpustart_t fn, void *arg)
```
Comment (dcpleung) [inline arch/x86/core/intel64/cpu.c:84]: Please do not mix formatting changes into commits containing functional changes, which makes it harder to review the functional changes. Ditto for formatting changes below.

**Review by dcpleung — DISMISSED** [review]

File: arch/x86/core/intel64/cpu.c [hunk 4]
```diff
@@ -155,10 +67,11 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 
 	apic_id = x86_cpu_loapics[cpu_num];
 
-	x86_cpuboot[cpu_num].sp = (uint64_t) Z_KERNEL_STACK_BUFFER(stack) + sz;
+	x86_cpuboot[cpu_num].sp = (uint64_t)Z_KERNEL_STACK_BUFFER(stack) + sz;
```
Comment (dcpleung) [inline arch/x86/core/intel64/cpu.c:106]: Please remove this change.

File: arch/x86/include/intel64/kernel_arch_data.h [hunk 5]
```diff
@@ -35,7 +36,46 @@ extern uint8_t x86_cpu_loapics[];	/* CPU logical ID -> local APIC ID */
 #endif /* _ASMLANGUAGE */
 
 #ifdef CONFIG_X86_KPTI
-#define Z_X86_TRAMPOLINE_STACK_SIZE	128
+#define Z_X86_TRAMPOLINE_STACK_SIZE 128
```
Comment (dcpleung) [inline arch/x86/include/intel64/kernel_arch_data.h:39]: Please remove this change.

File: arch/x86/include/intel64/kernel_arch_data.h (hunk 1 above)
Comment (dcpleung) [reply arch/x86/include/intel64/kernel_arch_data.h:29]: In that case, it would be a good idea to rename it to `apic_cpu_id` or something similar. The name `cpu_id` is used in subsequent commits where it is Zephyr's CPU ID and not APIC. Renaming it will avoid any confusion between two.

Comment (najumon1980) [reply arch/x86/include/intel64/kernel_arch_data.h]: sorry my mistake, this is cpu id pass down by kernel (not the HW CPU ID which is same as local APIC ID). Updated the comment

File: include/zephyr/kernel/thread.h [hunk 6]
```diff
@@ -121,7 +121,7 @@ struct _thread_base {
 
 #ifdef CONFIG_SCHED_CPU_MASK
 	/* "May run on" bits for each CPU */
-	uint8_t cpu_mask;
+	uint32_t cpu_mask;
```
Comment (tbursztyka) [inline include/zephyr/kernel/thread.h:124]: this is de-aligning the structure 

(SCHED_CPU_MASK does not depend on SMP, but check Kconfig help, it does not make sense to enable it without SMP)

File: arch/x86/include/intel64/kernel_arch_data.h [hunk 7]
```diff
@@ -38,4 +39,43 @@ extern uint8_t x86_cpu_loapics[];	/* CPU logical ID -> local APIC ID */
...
+#define ACPI_CPU_INIT(n, _)									\
+	uint8_t z_x86_exception_stack##n[CONFIG_X86_EXCEPTION_STACK_SIZE] __aligned(16);	\
+	uint8_t z_x86_nmi_stack##n[CONFIG_X86_EXCEPTION_STACK_SIZE] __aligned(16);		\
+	TRAMPOLINE_STACK(n);									\
+	Z_GENERIC_SECTION(.tss)									\
+	struct x86_tss64 tss##n = {								\
+		TRAMPOLINE_INIT(n)								\
+		.ist6 =	(uint64_t)z_x86_nmi_stack##n + CONFIG_X86_EXCEPTION_STACK_SIZE,		\
+		.ist7 = (uint64_t)z_x86_exception_stack##n + CONFIG_X86_EXCEPTION_STACK_SIZE,	\
+		.iomapb = 0xFFFF, .cpu = &(_kernel.cpus[n])}
```
Comment (tbursztyka) [inline arch/x86/include/intel64/kernel_arch_data.h:63]: have a new line for each attribute. Macro generated code should follow the same code style as the hand written one. (so have an empty line for the end } and properly indented as well.)

File: arch/x86/core/intel64/cpu.c [hunk 8]
```diff
@@ -108,7 +144,20 @@ FUNC_NORETURN void z_x86_cpu_init(struct x86_cpuboot *cpuboot)
...
+
 	/* Enter kernel, never return */
 	cpuboot->ready++;
 	cpuboot->fn(cpuboot->arg);
 }
+
+struct x86_cpu_info *z_x86_cpu_info_get(uint8_t cpu_id)
+{
+	if (cpu_id >= CONFIG_MP_MAX_NUM_CPUS) {
+		printk("invlaid cpu id:%d\n", cpu_id);
```
Comment (tbursztyka) [inline arch/x86/core/intel64/cpu.c:158]: typo

File: arch/x86/core/intel64/cpu.c [hunk 9]
```diff
@@ -168,15 +81,13 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
...
- */
 FUNC_NORETURN void z_x86_cpu_init(struct x86_cpuboot *cpuboot)
 {
+	uint8_t cpu_num;
+
 	x86_sse_init(NULL);
 
-	/* The internal cpu_number is the index to x86_cpuboot[] */
-	unsigned char cpu_num = (unsigned char)(cpuboot - x86_cpuboot);
+	cpu_num = cpuboot->cpu_id;
```
Comment (tbursztyka) [inline arch/x86/core/intel64/cpu.c:126]: since you move the variable declaration up to the code black start, then remove the next empty line (as if is testing the value of that assigned variable, which rule does not apply if the variable is declared and assigned at the same time)

Anyway, are you sure there is no misalignment of cpu_id and apic_id (as used in arch_start_cpu) ?

So far the cpu_num was not related to any hardware identifier, for the good reason that sometime there are CPUs with N cores but only K cores are actually valid and activated. But the inactivated cores can still be listed and owning a legit apic_id/cpu_id, and it's only through ACPI that these can be ignored (thus the logic behind acpi_local_apic_get(): it ignores the disabled ones).
Check this, make sure the cpu_id you are getting matches the same core which acip_id will be retrieved from ACPI, when there are inactive cores.

File: tests/arch/x86/hybrid/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (tbursztyka) [inline tests/arch/x86/hybrid/CMakeLists.txt:1]: commit title typo

File: tests/kernel/smp/src/main.c [hunk 11]
```diff
@@ -14,11 +14,12 @@
 #error SMP test requires at least two CPUs!
 #endif
 
+#define MAX_NUM_THREADS CONFIG_MP_MAX_NUM_CPUS
```
Comment (tbursztyka) [inline tests/kernel/smp/src/main.c:17]: A fix for ? You do not describe the actual issue the pre-defined value was generating. Describe it in your commit message

**Review by tbursztyka — DISMISSED** [review]

File: kernel/sched.c [hunk 12]
```diff
@@ -1621,8 +1621,8 @@ static inline int z_vrfy_k_is_preempt_thread(void)
 
 #ifdef CONFIG_SCHED_CPU_MASK
 # ifdef CONFIG_SMP
-/* Right now we use a single byte for this mask */
-BUILD_ASSERT(CONFIG_MP_MAX_NUM_CPUS <= 8, "Too many CPUs for mask word");
+/* Right now we use a single DWORD for this mask */
+BUILD_ASSERT(CONFIG_MP_MAX_NUM_CPUS <= 32, "Too many CPUs for mask word");
```
Comment (finikorg) [inline kernel/sched.c:1625]: I wonder do we need this assert at all? It is already handled through Kconfig, right?

File: arch/x86/core/intel64/cpu.c (hunk 8 above)
Comment (finikorg) [reply arch/x86/core/intel64/cpu.c:158]: should `printk()` be used here at all?

File: arch/x86/core/intel64/cpu.c [hunk 13]
```diff
@@ -108,7 +144,20 @@ FUNC_NORETURN void z_x86_cpu_init(struct x86_cpuboot *cpuboot)
 	z_x86_msr_write(X86_FMASK_MSR, EFLAGS_SYSCALL);
 #endif
 
+	/* Cache AP info. */
+	__ASSERT(!x86_update_cpu_info(cpu_num), "Get AP CPU info failed");
```
Comment (finikorg) [inline arch/x86/core/intel64/cpu.c:148]: Call this inside ASSERT is not the best way, are you sure ASSERT is enabled? Better check return code.

File: include/zephyr/arch/x86/intel64/structs.h [hunk 14]
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2023 Intel Corp.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_ARCH_X86_STRUCTS_H_
+#define ZEPHYR_INCLUDE_ARCH_X86_STRUCTS_H_
+
+enum x86_cpu_type {
+	cpu_type_atom = 0x20,  /* CPU type Intel Atom */
```
Comment (finikorg) [inline include/zephyr/arch/x86/intel64/structs.h:10]: Should enum type values be in capital?

File: boards/x86/intel_rpl/intel_rpl_s_crb_defconfig [hunk 15]
```diff
@@ -15,3 +15,4 @@ CONFIG_BUILD_OUTPUT_EFI=y
 CONFIG_BUILD_NO_GAP_FILL=y
 CONFIG_HEAP_MEM_POOL_SIZE=64000000
 CONFIG_MAIN_STACK_SIZE=320000
+CONFIG_MAX_THREAD_BYTES=11
```
Comment (finikorg) [inline boards/x86/intel_rpl/intel_rpl_s_crb_defconfig:18]: are you sure it is needed here?

File: tests/arch/x86/nmi/src/main.c (hunk 2 above)
Comment (finikorg) [reply tests/arch/x86/nmi/src/main.c:20]: Please also spell check word  `varible` in commit message and subject

File: tests/arch/x86/hybrid/src/smpboot.c [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+#include <zephyr/kernel.h>
+#include <zephyr/ztest.h>
+#include <zephyr/kernel_structs.h>
+
+BUILD_ASSERT(CONFIG_SMP);
+BUILD_ASSERT(CONFIG_X86_64);
+
+#define CPUID_MASK_WORD  0xFF
+#define CPUID_MASK_NIBLE 0x0F
+#define STACKSZ          2048
```
Comment (finikorg) [inline tests/arch/x86/hybrid/src/smpboot.c:15]: STACK_SIZE?

File: tests/arch/x86/hybrid/src/smpboot.c [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+
+volatile bool mp_flag[CONFIG_MP_MAX_NUM_CPUS];
+
+struct k_thread cpu_thr;
+K_THREAD_STACK_DEFINE(thr_stack, STACKSZ);
+
+extern void z_smp_start_cpu(int id);
+
+static int cpu_info_get(struct x86_cpu_info *info)
+{
```
Comment (finikorg) [inline tests/arch/x86/hybrid/src/smpboot.c:26]: I wonder could we test the function you wrote in `cpu.c` :  ` x86_update_cpu_info()`

File: tests/arch/x86/hybrid/testcase.yaml [hunk 18]
```diff
@@ -0,0 +1,9 @@
+tests:
+  arch.x86.hybrid:
+    arch_allow: x86
+    harness: console
+    harness_config:
+      type: one_line
+      regex:
+        - "hybrid: complete"
```
Comment (finikorg) [inline tests/arch/x86/hybrid/testcase.yaml:8]: I do not see that test would print this

**Review by finikorg — CHANGES_REQUESTED** [review]

File: tests/arch/x86/hybrid/src/smpboot.c [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	info->hybrid = (edx >> 15) & 0x01;
+
+	return 0;
+}
+
+static void print_cpu_info(struct x86_cpu_info *cpu_info)
+{
+	printk("CPU info[family:%x model:%x stepping:%x type:%s]\n", cpu_info->family,
+	       cpu_info->model, cpu_info->stepping, cpu_info->bsp ? "BSP" : "AP");
```
Comment (finikorg) [inline tests/arch/x86/hybrid/src/smpboot.c:55]: Where do you have `bsp` set?

Comment (finikorg) [reply tests/arch/x86/hybrid/src/smpboot.c:55]: Also BSP is a bad name

**Comment by finikorg** [comment]
@najumon1980 , I already commented that you put wrong check https://github.com/zephyrproject-rtos/zephyr/pull/60230#discussion_r1342478959. This result in grep for wrong string and the test is failing as you can see in the failing check section. Just remove it.

File: tests/arch/x86/hybrid/testcase.yaml (hunk 18 above)
Comment (najumon1980) [reply tests/arch/x86/hybrid/testcase.yaml]: updated

File: boards/x86/intel_rpl/intel_rpl_s_crb_defconfig [hunk 20]
```diff
@@ -13,3 +13,6 @@ CONFIG_X2APIC=y
 CONFIG_SMP=y
 CONFIG_BUILD_OUTPUT_EFI=y
 CONFIG_BUILD_NO_GAP_FILL=y
+CONFIG_HEAP_MEM_POOL_SIZE=64000000
+CONFIG_MAIN_STACK_SIZE=320000
```
Comment (jhedberg) [inline boards/x86/intel_rpl/intel_rpl_s_crb_defconfig:17]: It's quite odd to see the main stack explicitly set by the board - the requirements should pretty much 100% depend on the application based on what it does through its `main()` function. Why does it need to be set here instead of in the application?

Comment (najumon1980) [reply boards/x86/intel_rpl/intel_rpl_s_crb_defconfig]: Removed

Comment (finikorg) [reply boards/x86/intel_rpl/intel_rpl_s_crb_defconfig]: Don't we have set it already?

https://github.com/zephyrproject-rtos/zephyr/blob/f69641f7d204864aa26f8bdd9fecab259e535da2/boards/x86/intel_rpl/Kconfig.defconfig#L33-L37

Comment (jhedberg) [reply boards/x86/intel_rpl/intel_rpl_s_crb_defconfig]: @finikorg yes, I don't think any board should be doing that. The requirements for the `main()` function should be only set by the app. `Kconfig.defconfig` is at least a little bit better than `<board>_defconfig` since it's just modifying the default value, i.e. the app can still override it.

File: arch/x86/core/intel64/cpu.c [hunk 21]
```diff
@@ -12,126 +14,73 @@
...
-extern uint8_t z_x86_exception_stack1[];
-extern uint8_t z_x86_exception_stack2[];
-extern uint8_t z_x86_exception_stack3[];
+#if !defined(CONFIG_ACPI)
+#define INIT_CPUID(n, _) n
+__weak uint8_t x86_cpu_loapics[] = {
+	LISTIFY(CONFIG_MP_MAX_NUM_CPUS, INIT_CPUID, (,)),};
+#else
+__weak uint8_t x86_cpu_loapics[CONFIG_MP_MAX_NUM_CPUS];
+#endif
```
[the remark this answers is not in the document]
Comment (najumon1980) [reply arch/x86/core/intel64/cpu.c]: updated

File: kernel/Kconfig [hunk 22]
```diff
@@ -1168,14 +1168,14 @@ config SMP_BOOT_DELAY
 config MP_NUM_CPUS
 	int "Number of CPUs/cores [DEPRECATED]"
 	default MP_MAX_NUM_CPUS
-	range 1 12
+	range 1 32
```
Comment (andyross) [inline kernel/Kconfig:1171]: Pedantry: this is technically a bisection problem, as you're increasing the range in kconfig before adding the code to handle it.  Properly this patch should be last.  I doubt anyone cares.

File: kernel/sched.c (hunk 12 above)
Comment (andyross) [reply kernel/sched.c:1730]: +1 to removal, this is no doubt descended from some early code without a limit in kconfig.

File: arch/x86/core/intel64/locore.S [hunk 23]
```diff
@@ -874,6 +834,50 @@ IRQ(232); IRQ(233); IRQ(234); IRQ(235); IRQ(236); IRQ(237); IRQ(238); IRQ(239)
 IRQ(240); IRQ(241); IRQ(242); IRQ(243); IRQ(244); IRQ(245); IRQ(246); IRQ(247)
 IRQ(248); IRQ(249); IRQ(250); IRQ(251); IRQ(252); IRQ(253); IRQ(254); IRQ(255)
 
+.code16
+.align 4096
+.global x86_ap_start
+x86_ap_start:
```
Comment (andyross) [inline arch/x86/core/intel64/locore.S:840]: It's unclear why this function is moving, it looks unchanged to me? (Or if it is changed, the deltas are being hidden by code motion)  Can you rework the patch so that the move isn't present?  Or if it does have to move can you put the "motion only" into a separate refactor patch?

Basically: I'm terrified there's something subtly different between two blocks of very subtle assembly code, and the structure of this patch makes that impossible to see.

File: arch/x86/core/intel64/cpu.c [hunk 24]
```diff
@@ -46,6 +46,42 @@ struct x86_cpuboot x86_cpuboot[] = {
 	LISTIFY(CONFIG_MP_MAX_NUM_CPUS, X86_CPU_BOOT_INIT, (,)),
 };
 
+int x86_update_cpu_info(uint8_t cpu_id)
```
Comment (andyross) [inline arch/x86/core/intel64/cpu.c:49]: I don't see why this patch is in this PR?  It doesn't have anything to do with CPU count.  Can you do this as a separate submission?

File: boards/x86/intel_rpl/intel_rpl_s_crb_defconfig (hunk 15 above)
Comment (andyross) [reply boards/x86/intel_rpl/intel_rpl_s_crb_defconfig:16]: That's counting total threads, not CPUs (it allows for 88 threads total).  You're saying the per-CPU idle threads (the only ones that scale with core count) are pushing it across the limit?  Please clarify in the patch description.  The default value is "2", so this is adding space for 72 new threads, not the "six more rounded up to a multiple of 8" I would expect. I agree this looks suspicious.

File: tests/arch/x86/nmi/src/main.c (hunk 2 above)
Comment (andyross) [reply tests/arch/x86/nmi/src/main.c:20]: Still don't get why you've included a whole patch in the PR just to glue a "0" onto the end of a variable that seems perfectly clear on its own?   Not a big deal, just weird.

File: tests/arch/x86/hybrid/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,8 @@
...
+# Copyright (c) 2023 Intel Corp.
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(intel_hybrid)
```
Comment (andyross) [inline tests/arch/x86/hybrid/CMakeLists.txt:6]: As mentioned elsewhere, I'd really prefer to see "Intel hybrid CPU support" be a separate PR instead of mixed with arch-level changes.

File: tests/kernel/smp/src/main.c [hunk 26]
```diff
@@ -14,11 +14,12 @@
...
 #define RUN_FACTOR (CONFIG_SMP_TEST_RUN_FACTOR / 100.0)
 
 #define T2_STACK_SIZE (2048 + CONFIG_TEST_EXTRA_STACK_SIZE)
 #define STACK_SIZE (384 + CONFIG_TEST_EXTRA_STACK_SIZE)
-#define DELAY_US 50000
+#define DELAY_US (50000 * MAX_NUM_THREADS)
```
Comment (andyross) [inline tests/kernel/smp/src/main.c:22]: This seems suspect.  50ms is a **very long time** already.  Surely it doesn't take 50ms per core to do anything in this test.  Would prefer to see whatever this failure is root caused.  This looks like a hack to me.

**Review by andyross — COMMENTED** [review]
This came by again and reminded me I never reviewed.  Some notes.  None rise to a -1 on their own I don't think, but would be good to see answered/addressed.

File: tests/kernel/smp/src/main.c (hunk 26 above)
Comment (najumon1980) [reply tests/kernel/smp/src/main.c]: Removed the changes

**Comment by najumon1980** [comment]
@andyross @jhedberg  could you please review the code ?

**Comment by najumon1980** [comment]
@andyross @jhedberg could you please review the code so that it can be part of next release ?

**Comment by najumon1980** [comment]
@andyross @tbursztyka @jhedberg could you please approve if no more comments ?

**Comment by najumon1980** [comment]
@andyross @tbursztyka @jhedberg could you please approve if no more comments ?

File: kernel/sched.c (hunk 12 above)
Comment (TaiJuWu) [reply kernel/sched.c:1739]: Build_ASSERT is checked in compile time, right?
Is there any drawback to keep it?

Comment (jhedberg) [reply kernel/sched.c:1739]: > Build_ASSERT is checked in compile time, right? Is there any drawback to keep it?

Unnecessary code & redundancy. It also gives the false impression that the bounds are not enforced anywhere else even though they are (in Kconfig).

**Comment by najumon1980** [comment]
> > @andyross @tbursztyka @jhedberg could you please approve if no more comments ?
> 
> @najumon1980 there still seems to be several unresolved comments, like these: [#60230 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/60230#discussion_r1424132317) [#60230 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/60230#discussion_r1424143033) [#60230 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/60230#discussion_r1424130381)
> 
> Once these are addressed it'd also be important to get another look from @andyross

updated and kindly request you to review

**Comment by jhedberg** [comment]
@najumon1980 your first commit seems to be empty ("kernel: remove redundant check for number of cpu"). Seems like it should just be removed, or is there some other mistake?

**Comment by najumon1980** [comment]
> BUILD_ASSERT(CONFIG_MP_MAX_NUM_CPUS <



> @najumon1980 your first commit seems to be empty ("kernel: remove redundant check for number of cpu"). Seems like it should just be removed, or is there some other mistake?

the latest mainline code already removed BUILD_ASSERT(CONFIG_MP_MAX_NUM_CPUS < and hence this commit no more required.

**Review by jhedberg — DISMISSED** [review]

**Comment by najumon1980** [comment]
@andyross @tbursztyka could you please approve if no more comments ?

**Review by cfriedt — DISMISSED** [review]

**Review by finikorg — DISMISSED** [review]

**Review by peter-mitsis — DISMISSED** [review]

**Comment by najumon1980** [comment]
@jhedberg  @cfriedt @finikorg @peter-mitsis , just done a rebase to fix merge conflict and could you please approve once again ?

**Review by cfriedt — APPROVED** [review]

**Review by peter-mitsis — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by finikorg — APPROVED** [review]

**Review by andyross — APPROVED** [review]
Trying to stay out of the x86 arch business these days, but everything looks clean to me.

**Comment by najumon1980** [comment]
@dcpleung could you review and approve from your side

**Review by dcpleung — APPROVED** [review]
